cpp/src/reductions/scan.cu

```diff
@@ -2,28 +2,88 @@
 #include "rmm/rmm.h"
 #include "utilities/cudf_utils.h"
 #include "utilities/error_utils.h"
+#include "bitmask/bit_mask.cuh"
+#include "utilities/type_dispatcher.hpp"
 
 #include <cub/device/device_scan.cuh>
 
 
+template <class T>
+__global__
+void gpu_copy_mask(const T *data, const gdf_valid_type *mask,
+    gdf_size_type size, T *results, T identity)
+{
+    int id = blockIdx.x * blockDim.x + threadIdx.x;
+    if (id >= size)return;
+
+    const BitMask bitmask(reinterpret_cast<bit_mask_t*>(const_cast<gdf_valid_type*>(mask)), size);
+    bool presence = bitmask.is_valid(id);
+    results[id] = (presence) ? data[id] : identity;
+}
+
+#define COPYMASK_BLOCK_SIZE 1024
+
+template <class T>
+inline
+void apply_copy_mask(const T *data, const gdf_valid_type *mask,
+    gdf_size_type size, T *results, T identity)
+{
+    int blocksize = (size < COPYMASK_BLOCK_SIZE ?
+        size : COPYMASK_BLOCK_SIZE);
+    int gridsize = (size + COPYMASK_BLOCK_SIZE - 1) /
+        COPYMASK_BLOCK_SIZE;
+
+    // launch kernel
+    gpu_copy_mask << <gridsize, blocksize >> > (
+        data, mask, size, results, identity);
+}
+
 
 template <class T>
 struct Scan {
     static
-    gdf_error call(const T *inp, T *out, size_t size, bool inclusive) {
+    gdf_error call(const gdf_column *inp, gdf_column *out, bool inclusive) {
         using cub::DeviceScan;
-
+        cudaStream_t stream = 0;  // TODO: non-default stream
         auto scan_function = (inclusive? inclusive_sum : exclusive_sum);
+        size_t size = inp->size;
+        const T* input = reinterpret_cast<const T*>(inp->data);
+        T* output = reinterpret_cast<T*>(out->data);
 
         // Prepare temp storage
         void *temp_storage = NULL;
         size_t temp_storage_bytes = 0;
-        scan_function(temp_storage, temp_storage_bytes, inp, out, size);
-        RMM_TRY( RMM_ALLOC(&temp_storage, temp_storage_bytes, 0) ); // TODO: non-default stream
-        // Do scan
-        scan_function(temp_storage, temp_storage_bytes, inp, out, size);
+        scan_function(temp_storage, temp_storage_bytes, input, output, size);
+        RMM_TRY( RMM_ALLOC(&temp_storage, temp_storage_bytes, stream) );
+
+        if (inp->valid) {
+            // copy bitmask
+            size_t valid_byte_length = (size + 7) / 8;
+            cudaError_t error = cudaMemcpyAsync(out->valid, inp->valid,
+                valid_byte_length, cudaMemcpyDeviceToDevice, stream);
+            out->null_count = inp->null_count;
+
+            T* temp_inp;    
+            RMM_TRY(RMM_ALLOC(&temp_inp, size * sizeof(T), stream));
+            // copy input data and replace with 0 if mask is not valid
+            apply_copy_mask(reinterpret_cast<const T*>(inp->data), inp->valid,
```

Comment:
Instead of re-implementing this functionality, use this https://github.com/rapidsai/cudf/pull/643/files#diff-3336b1441d3b2bb3e4d27ca7a8482146R976 (note that this is still a pending PR, you can merge it into your branch, we'll just need to make sure we merge the other PR before yours is merged.)

Reply:
What this PL wants is to copy input into output and then replace nulls with scalar value(0) at once.
replace_nulls_with_scalar() only replaces nulls in output column.
Is it possible to add copy_and_replace_nulls_with_scalar() API in PL #643 ?
Or, if it is legal to replacing nulls in input column with a scalar value, it is mush simple. It is legal in Apache Arrow spec, but it breaks if another function do simultenously replacing operation with different scalar value.

Reply:
Couldn't you just copy the `input` column, and then call `replace_nulls` on the copy? 

Technically you could do the copy and replace the null values in a single pass, but I'm not convinced it would matter.

Reply:
Though I've implemented using `replace_nulls`, the implementation looks redundant to me since gdf_replace_nulls requires gdf_column as input.
Here is the implementation, do we go with this?

```
{   // instead of calling apply_copy_mask
    // create a temporary column for gdf_replace_nulls
    gdf_column temp_column;
    gdf_error gdf_column_view(&temp_column,
        temp_input, input->valid, input->size, input->dtype);

    // copy input data into result
    CUDA_TRY(cudaMemcpyAsync(temp_input, d_input,
        input->size * sizeof(T), cudaMemcpyDeviceToDevice, stream));

    // create a schalar storage and put the identity value
    T* scholar_storage = NULL;
    RMM_TRY(RMM_ALLOC(&scholar_storage, sizeof(T), stream));
    T scholar_value(0);
    CUDA_TRY(cudaMemcpyAsync(scholar_storage, &scholar_value,
        sizeof(T), cudaMemcpyHostToDevice, stream));

    // create a schalar column for gdf_replace_nulls
    gdf_column scholar_column;
    gdf_error gdf_column_view(&scholar_column,
        schalar_storage, NULL, 1, results->dtype);

    // replace nulls in result with schalar value in scholar_column
    gdf_replace_nulls(&temp_column, &scholar_column);

    RMM_TRY(RMM_FREE(scholar_storage, stream));
}
```

Reply:
@kovaltan : (Not an authoritative statement) I think you can adapt the implementation of gdf_replace_nulls so that its current interface just wraps an interface which satisfies your needs as well, e.g. works on raw data rather than a `gdf_column`.

Reply:
At offline discussion with Jake, we won't use `replace_nulls` for now.